File: app/src/main/java/edu/luc/etl/cs313/misc/boundedcounter/model/SimpleBoundedCounter.java
package edu.luc.etl.cs313.misc.boundedcounter.model;

import edu.luc.etl.cs313.android.clickcounter.BuildConfig;
import edu.luc.etl.cs313.misc.boundedcounter.cli.BoundedCounter;

/**
 * A simple bounded counter implementation.
 */
public class SimpleBoundedCounter implements BoundedCounter {

    /** The lower bound of the counter. */
    private final int min;

    /** The upper bound of the counter. */
    private final int max;

    /** The current value of the counter. */
    private int value;

    /** Constructs a bounded counter with the default bounds. */
    public SimpleBoundedCounter() {
        this(0, 10);
    }

    /**
     * Constructs a bounded counter with the given bounds.
     *
     * @param min the lower bound
     * @param max the upper bound
     * @throws IllegalArgumentException if min > max
     */
    public SimpleBoundedCounter(final int min, final int max) {
        if (min >= max) {
            throw new IllegalArgumentException("min >= max");
        }
        this.min = min;
        this.max = max;
        this.value = this.min;
    }

    /**
     * Indicates whether this counter currently satisfies its internal data
     * invariant: the counter value is within the bounds.
     *
     * @return whether the data invariant is satisfied
     */
    protected boolean dataInvariant() {
        return min <= value && value <= max;
    }

    @Override
    public void increment() {
        assertIfDebug(() -> dataInvariant() && !isFull());
        ++value;
        assertIfDebug(() -> dataInvariant());
    }

    @Override
    public void decrement() {
        assertIfDebug(() -> dataInvariant() && !isEmpty());
        --value;
        assertIfDebug(() -> dataInvariant());
    }

    @Override
    public int get() {
        return value;
    }

    @Override
    public boolean isFull() {
        return value >= max;
    }

    @Override
    public boolean isEmpty() {
        return value <= min;
    }

    public interface BooleanSupplier {
        boolean getAsBoolean();
    }

    protected void assertIfDebug(final BooleanSupplier p) {
        if (BuildConfig.DEBUG && !p.getAsBoolean()) {
            throw new AssertionError();
        }
    }
}

File: app/src/main/kotlin/edu/luc/etl/cs313/misc/boundedcounter/model/SimpleBoundedCounter.kt
package edu.luc.etl.cs313.misc.boundedcounter.model

import edu.luc.etl.cs313.android.clickcounter.BuildConfig
import edu.luc.etl.cs313.misc.boundedcounter.cli.BoundedCounter

/**
 * A simple bounded counter implementation.
 */
class SimpleBoundedCounter @JvmOverloads constructor(val min: Int = 0, val max: Int = 10) : BoundedCounter {
/**
 * Constructs a bounded counter with the given bounds.

 * @param min the lower bound
 * *
 * @param max the upper bound
 * *
 * @throws IllegalArgumentException if min > max
 */

    /** The current value of the counter.  */
    private var value: Int = 0

    init {
        if (min >= max) {
            throw IllegalArgumentException("min >= max")
        }
        this.value = this.min
    }

    /**
     * Indicates whether this counter currently satisfies its internal data
     * invariant: the counter value is within the bounds.

     * @return whether the data invariant is satisfied
     */
    protected fun dataInvariant(): Boolean {
        return min <= value && value <= max
    }

    override fun increment() {
        assertIfDebug({ dataInvariant() && !isFull })
        ++value
        assertIfDebug({ this.dataInvariant() })
    }

    override fun decrement() {
        assertIfDebug({ dataInvariant() && !isEmpty })
        --value
        assertIfDebug({ this.dataInvariant() })
    }

    override fun get(): Int {
        return value
    }

    override val isFull: Boolean
        get() = value >= max

    override val isEmpty: Boolean
        get() = value <= min

    protected fun assertIfDebug(p: () -> Boolean) {
        if (BuildConfig.DEBUG && !p()) {
            throw AssertionError()
        }
    }
}
/** Constructs a bounded counter with the default bounds.  */
